server_stubs.go

```diff
@@ -1,32 +1,51 @@
-// +build !cgo plan9 windows android
+// +build !cgo windows android
 
 package sftp
 
 import (
 	"fmt"
 	"os"
+	"strconv"
 	"time"
+
+	sshfx "github.com/pkg/sftp/internal/encoding/ssh/filexfer"
 )
 
+func formatID(id uint32) string {
+	return strconv.FormatUint(uint64(id), 10)
+}
+
+// ls -l style output for a file, which is in the 'long output' section of a readdir response packet
+// this is a very simple (lazy) implementation, just enough to look almost like openssh in a few basic cases
 func runLs(dirname string, dirent os.FileInfo) string {
-	typeword := runLsTypeWord(dirent)
-	numLinks := 1
-	if dirent.IsDir() {
-		numLinks = 0
+	// example from openssh sftp server:
+	// crw-rw-rw-    1 root     wheel           0 Jul 31 20:52 ttyvd
+	// format:
+	// {directory / char device / etc}{rwxrwxrwx}  {number of links} owner group size month day [time (this year) | year (otherwise)] name
+
+	symPerms := fromFileMode(dirent.Mode()).String()
+
+	var numLinks uint64 = 1
+	uid, gid := "0", "0"
+
+	switch sys := dirent.Sys().(type) {
+	case *sshfx.Attributes:
+		uid = formatID(sys.UID)
+		gid = formatID(sys.GID)
+	case *FileStat:
+		uid = formatID(sys.UID)
+		gid = formatID(sys.GID)
 	}
-	username := "root"
-	groupname := "root"
+
 	mtime := dirent.ModTime()
-	monthStr := mtime.Month().String()[0:3]
-	day := mtime.Day()
-	year := mtime.Year()
-	now := time.Now()
-	isOld := mtime.Before(now.Add(-time.Hour * 24 * 365 / 2))
-
-	yearOrTime := fmt.Sprintf("%02d:%02d", mtime.Hour(), mtime.Minute())
-	if isOld {
-		yearOrTime = fmt.Sprintf("%d", year)
+	date := mtime.Format("Jan 2")
+
+	var yearOrTime string
+	if mtime.Before(time.Now().AddDate(0, -6, 0)) {
+		yearOrTime = mtime.Format("2006")
+	} else {
+		yearOrTime = mtime.Format("15:04")
 	}
```

Comment:
It occurred to me that we were doing a lot of manual fiddling with formatting time values, when we could just be using the formating that is made available to us.